Add Depends to monero

TheCharlatan opened the pull request:
Depends is a localized cross compilation environment originally developed by Cory Fields. It allows its users to create cross compiled binaries for 32 and 64 bit linux and windows, arm hf, aarch64 and macos.

This pull request places depends into contrib/depends. The required monero dependencies are generated for each platform triplet (for example `x86_64-w64-mingw32`) and saved in their own directory in depends. Recipes for each package can be found in `contrib/depends/packages`, `packages.mk` defines the packages used for each platform triplet. The allowed ${triplet}'s are: `i686-w64-mingw32`, `x86_64-w64-mingw32`, `x86_64-apple-darwin11`, `x86_64-linux-gnu`, `i686-linux-gnu`, `arm-linux-gnueabihf` and `aarch64-linux-gnu`.

Libraries and tools are generated in depends by running `make HOST=${triplet}`. This step first downloads tar archives of the required dependency source files and places them in sources/. These archives are then checked for their sha256sum hash and compared with a hardcoded value in the package recipe.  They are then extracted and placed into work/build during a build process. After building the package binaries/libraries are installed in work/staging . Once completed their staging directory is compressed and placed into built/ , after which their content in work/ is deleted. When every package is in built/ their archives are decompressed into ${triple}/ . This is where the monero build system will pick them up from. 

A cmake toolchain file is uniquely generated for ${triple} and placed in `contrib/depends/${triplet}/share/toolchain.cmake`. They can then be accessed by the monero build system by executing for example: ``cmake -DCMAKE_TOOLCHAIN_FILE=`pwd`/contrib/depends/${triplet}/share/toolchain.cmake`` . The toolchain file also sets the DEPENDS variable to identify a depends build to the monero build system. This is used to exclude some additions from external as well as correcting some make flags.

General usage is also briefly reflected in the Readme.md under "Building portable statically linked binaries".

### Example of a depends build for 64 bit windows: 

Install build essentials, nsis and g++-mingw-w64-x86-64 on your host system. Then:
```
    cd contrib/depends
    make HOST=x86_64-w64-mingw32
```
Once it is finished building go back to the root directory and execute: 
```
    cmake -DCMAKE_TOOLCHAIN_FILE=`pwd`/contrib/depends/x86_64-w64-mingw32/share/toolchain.cmake
```

### Advantages of using depends:
- Serves as a basis for reproducible builds on all plattforms
- Allows cross compilation of windows and macos monero binaries
- Having a standardized and well defined list of dependencies used in the project.
- Vendored dependencies can now be source controlled from their original repository 

One of the things changed that also affect builds other than with the depends system is lowercasing the winsock header include.
Another is that in this pull request the usage of unwind on windows is disable in the source code. Some discussion needs to be had about it. In its current form it only allows builds from depends.

Further topics that need to be discussed are the combinations of system and processor architecture pairs that need to be able to be passed to the monero build system for armv6, v7 and v8. 

I realize this is hard to review, and if there is anything I can do to make it easier, do not hesitate to leave suggestions. So far I tested binaries for linux and windows 64bit.

moneromooo-monero commented on .travis.yml:
```diff
@@ -0,0 +1,61 @@
+  - $HOME/.ccache
+env:
+  global:
+    - MAKEJOBS=-j3
```
That should not be overridden (or there should be a user override).

TheCharlatan replied:
This is a travis CI instruction set. Why is the override problematic? I only included it, because it is quite useful, but basically it does not much more than the current build bot.

moneromooo-monero replied:
Because building monero gobbles up a lot of RAM and 3 jobs means at least 6 GB.

TheCharlatan replied:
Removed MAKEJOBS.

moneromooo-monero commented on cmake/FindLibunwind.cmake:
```diff
@@ -18,9 +18,10 @@ find_path(LIBUNWIND_INCLUDE_DIR libunwind.h
 find_library(LIBUNWIND_LIBRARIES NAMES unwind )
 if(NOT LIBUNWIND_LIBRARIES STREQUAL "LIBUNWIND_LIBRARIES-NOTFOUND")
-  if (CMAKE_COMPILER_IS_GNUCC)
+  if (CMAKE_COMPILER_IS_GNUCC AND NOT DEPENDS)
```
That might be better off done in the caller, since it's find macro ?

TheCharlatan replied:
This was actually not needed anymore. Removed.

moneromooo-monero commented on src/common/stack_trace.cpp:
```diff
@@ -119,7 +119,6 @@ void log_stack_trace(const char *msg)
   int status;
   const char *log = stack_trace_log.empty() ? NULL : stack_trace_log.c_str();
 #endif
-
```
Please remove :)

TheCharlatan commented:
I removed the travis yml from the pr. A lot of packages are required (like gcc > v5) that are not available on the standard travis distro (ubuntu trusty). This could be solved by starting a docker container, that then runs a more up to date distro, but this goes over the scope of this pr. In general using travis with depends should be encouraged, since it allows cross platform CI builds on all repository forks.

moneromooo-monero commented:
Is it worth removing it just to fix the -j setting ?

TheCharlatan commented:
It's not just the -j setting. The entire CI needs to be able to run in a docker container, since the current travis distro is out of date (cmake and gcc are too old).

TheCharlatan commented:
Rebased. I removed the depends recipes for unbound and miniupnp, instead the build flags are passed to the submodules. Windows builds will work, as soon as the unbound cmake does not check for the winpthread dll anymore.

TheCharlatan commented:
Opened (#3)[https://github.com/monero-project/unbound/pull/3] on the unbound fork to remove the winpthread check.

erciccione commented:
HI @TheCharlatan could you give an ETA for this PR?

TheCharlatan commented:
@erciccione still working on it. Currently the following issues are still remaining, and the reasons why my [local travis builds](https://travis-ci.org/TheCharlatan/monero/builds/401949880?utm_source=github_status&utm_medium=notification) are failing:

1. Pcsc is not easily statically cross compiled. While the headers are available in both mingw (for windows) and the macOS SDK, the actual library does not seem to be there. Pcsc-lite itself does not support mac cross compilation as is, so I am a bit stuck there. I will probably take some time end of July to work on this by dropping pcsc and instead use hid for the hardware communication. I already talked to the ledger devs about this and they all agreed to it. 
2. This one should be easy, but I can't seem to figure out why the arm build does not want to shadow build anymore. If you look at the travis logs I posted above, cmake does not find any of the included subdirectories.
3. A bunch of smaller stuff, like inconsistent linking of icu etc. is also still in, but should be easy / can be fixed later.

moneromooo-monero commented:
It would be very, very nice if we could have this in time for the 0.13 builds at least for x86_64.
Since libpcsc is being replaced, that just leaves the icu stuff. Any chance this can be done soon ?

TheCharlatan commented:
Rebased. 
Current state looks like this (ticks indicate that the pr behaves as intended):
- [x] Linux x86_64 (looks like the compilation problem I am getting here is from master, it builds the daemon fine)
- [ ] 32-bit arm (compiler option not chosen correctly)
- [ ] 64-bit arm (need source for librt)
- [ ] 64-bit windows (needs hid)
- [ ] 64-bit mac (needs hid)

moneromooo-monero commented on CMakeLists.txt:
```diff
@@ -436,9 +440,19 @@ add_definition_if_function_found(strptime HAVE_STRPTIME)
+  STAMP_DIR ${LRELEASE_PATH}
+  CMAKE_ARGS -DLRELEASE_PATH=${LRELEASE_PATH}
+  INSTALL_COMMAND cmake -E echo "")
+#>>>>>>> b1f43170... Add lrelease to the depends
```
Conflict markers

moneromooo-monero commented on CMakeLists.txt:
```diff
@@ -812,8 +832,12 @@ endif()
   set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -Wa,-mbig-obj")
-  set(EXTRA_LIBRARIES mswsock;ws2_32;iphlpapi;crypt32)
-  set(ICU_LIBRARIES ${Boost_LOCALE_LIBRARY} icuio icuin icuuc icudt icutu iconv)
+  set(EXTRA_LIBRARIES mswsock;ws2_32;iphlpapi)
```
crypt32 ?

iDunk5400 replied:
crypt32 is needed on Windows, yes.

moneromooo-monero commented on contrib/depends/Makefile:
```diff
@@ -0,0 +1,227 @@
+NO_QT ?=
+NO_WALLET ?=
+NO_UPNP ?=
+FALLBACK_DOWNLOAD_PATH ?= https://bitcoincore.org/depends-sources
```
Does this mean the build process actually downloads stuff ? If it does, does it actually execute it ?

TheCharlatan replied:
getmonero should make the sources available as well, so we can change the download path from bitcoincore.org to getmonero.org .

moneromooo-monero commented on contrib/depends/description.md:
```diff
@@ -0,0 +1,53 @@
+Each package is configured and patched so that it will yield the same
+build-results with each consequent build, within a reasonable set of
+constraints. Some things like timestamp insertion are unavoidable, and are
+beyond the scope of this system. Additionally, the toolchain itself must be
```
Does this not defeat the point or being able to check a hash, rather than diff binaries and work out that all differences are down to timestamps (which may be pretty hard for a non expert) ?

moneromooo-monero commented on contrib/depends/description.md:
```diff
@@ -0,0 +1,53 @@
+
+Each package must define its source location and checksum. The build will fail
+if the fetched source does not match. Sources may be pre-seeded and/or cached
+as desired.
```
That probably answers my question above, seems good.

iDunk5400 commented:
For 64-bit Windows, you need mingw-w64-x86_64-hidapi in MSYS2.

iDunk5400 commented:
Nevermind my last comment, it was wrong. Looks like MSYS2 also uses libhidapi from [here](https://github.com/signal11/hidapi).

Technohacker commented on README.md:
````diff
@@ -532,6 +532,22 @@ Installing a snap is very quick. Snaps are secure. They are isolated with all of
+The required packages are the names for each toolchain on apt. Depending on your distro, they may have different names.
+Then go back to the source dir and type for example for windows 64bit: 
+
+* ```cmake -DCMAKE_TOOLCHAIN_FILE=`pwd`/contrib/depends/x86_64-w64-mingw32```
````
This part seems incomplete, I think it's supposed to be `cmake -DCMAKE_TOOLCHAIN_FILE=`pwd`/contrib/depends/x86_64-w64-mingw32/share/toolchain.cmake`

Technohacker replied:
I'll get it done tomorrow :)
